File: lib/data/episode.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/foundation.dart';
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:intl/intl.dart';
import 'package:podcast/data/firebase_converters/date_time_converter.dart';
import 'package:podcast/data/firebase_converters/duration_converter.dart';
import 'package:podcast/data/podcast.dart';
import 'package:podcast/extensions/nullability_extensions.dart';
import 'package:podcast/extensions/response_extension.dart';
import 'package:podcast/extensions/xml_element_extension.dart';
import 'package:podcast/services/podcast_audio_handler.dart';
import 'package:xml/xml.dart';

part 'episode.freezed.dart';
part 'episode.g.dart';

@freezed
class Episode with _$Episode implements ToJson {
  const factory Episode({
    required String guid,
    required Uri url,
    required String title,
    @DateTimeConverter() required DateTime? pubDate,
    required String? description,
    required Uri imageUrl,
    @DurationConverter() required Duration? duration,
    // region Dynamic fields
    @Default(false) bool listened,
    @DurationConverter() Duration? currentPosition,
    // endregion Dynamic fields
  }) = _Episode;

  const Episode._();

  factory Episode.fromJson(Map<String, dynamic> json) =>
      _$EpisodeFromJson(json);

  PodcastMediaItem mediaItem(DocumentSnapshot<Episode> episodeSnapshot) =>
      PodcastMediaItem(
        episode: episodeSnapshot,
        id: url.toString(),
        title: title,
        artUri: imageUrl,
        duration: duration,
      );

  /// Adds the fields from [other] that are set at runtime, such as [listened]
  /// and [currentPosition].
  Episode operator +(Episode other) => copyWith(
        listened: other.listened,
        currentPosition: other.currentPosition,
      );

  static final _rfc822Format = DateFormat('EEE, dd MMM yyyy HH:mm:ss');

  static Iterable<Episode> fromXml(
    XmlDocument document, {
    required Podcast podcast,
  }) sync* {
    for (final item in document.findAllElements('item')) {
      // region Required fields
      final guid = Uri.encodeComponent(item.getElementContent('guid')!);
      if (guid.contains('/')) {
        throw UnsupportedError(item.toString());
      }

      final pubDateValue = item.getElementContent('pubDate')!;
      final pubDate =
          _rfc822Format.tryParse(pubDateValue) ?? DateTime.parse(pubDateValue);

      final link = item.getElement('enclosure')?.getAttribute('url');
      if (link == null) {
        debugPrint('Link null for: $item');
      }
      final title = item.getElementContent('title') ??
          item.getElementContent('itunes:title');
      if (title == null) {
        debugPrint('Title null for: $item');
      }
      final durationString = item.getElementContent('itunes:duration');
      final duration =
          switch (durationString?.split(':').map(int.parse).toList()) {
        null => null,
        [final hours, final minutes, final seconds] => Duration(
            hours: hours,
            minutes: minutes,
            seconds: seconds,
          ),
        [final minutes, final seconds] => Duration(
            minutes: minutes,
            seconds: seconds,
          ),
        [final seconds] => Duration(seconds: seconds),
        _ => null,
      };
      // endregion Required fields

      // region Optional fields
      final imageUrl = item.getElement('itunes:image')?.getAttribute('href');
      final description = item.getElementContent('description');
      // endregion Required fields

      yield Episode(
        guid: guid,
        url: Uri.parse(link!),
        title: title!,
        duration: duration,
        pubDate: pubDate,
        description: description,
        imageUrl: imageUrl?.let(Uri.parse) ?? podcast.image,
      );
    }
  }
}

File: supabase/functions/_shared/data/Episode.ts
export class Episode {
    id!: string;
    url!: URL;
    title!: string;
    pub_date?: Date;
    description?: string;
    image_url!: URL;
    duration?: number;
    podcast_id!: string;
    constructor(
        options: {
            id: string;
            url: URL;
            title: string;
            pubDate: string;
            description: string;
            imageUrl: URL;
            duration: string;
            podcast_id: string;
        },
    ) {
        this.id = (options.podcast_id + ":" + options.id)!;
        this.url = options.url!;
        this.title = options.title!;
        this.pub_date = new Date(options.pubDate);
        this.description = options.description;
        this.image_url = options.imageUrl!;
        this.duration = parseDuration(options.duration);
        this.podcast_id = options.podcast_id!;

        // Fail if required fields are missing
        if (this.id === undefined) {
            throw new Error("Episode ID is required");
        } else if (this.url === undefined) {
            throw new Error("Episode URL is required");
        } else if (this.title === undefined) {
            throw new Error("Episode title is required");
        } else if (this.image_url === undefined) {
            throw new Error("Episode image URL is required");
        } else if (this.podcast_id === undefined) {
            throw new Error("Episode podcast ID is required");
        }
    }
}

function parseDuration(durationString: string): number | undefined {
    if (durationString === undefined) {
        return undefined;
    }

    // Try converting to number
    if (!isNaN(Number(durationString))) {
        return Number(durationString);
    }

    const parts = durationString.split(":").map(Number);

    let hours: number, minutes: number, seconds: number;
    switch (parts.length) {
        case 3:
            [hours, minutes, seconds] = parts;
            return (hours * 3600 + minutes * 60 + seconds);
        case 2:
            [minutes, seconds] = parts;
            return (minutes * 60 + seconds);
        case 1:
            [seconds] = parts;
            return seconds;
        default:
            return undefined; // Invalid format
    }
}
